### PR TITLE
Set `:shell_history` after formatting application partition on first boot

### PR DESCRIPTION
**Pull request by jjcarstens**
~~If we formatted and it successfully mounted, we need to restart before using the partition to prevent errors. This typically only occurs on first boot~~

Edit (after much discussion below):
After a fresh SD card burn, the application partition may not be
formatted which means the shell history will fail to start preventing
history being saved until a reboot.

This change detects if the application partition needed to be formatted
and resets the shell history env var to enable history again once
the disk was successfully formatted and mounted. This will allow
history to be saved even on first boot.

### CODE REVIEW

**Comment by ConnorRigby** [comment]
Does this work on OTP 24?

**Review by mobileoverlord — APPROVED** [review]
Does this also enable shell history, or is that only on a full vm restart?

**Comment by jjcarstens** [comment]
Yes, the intent is that the full VM restart enables the shell history since it will then allow this newly formatted app partition to be writable. This also seems to be the pattern that raspbian uses as well (First boot gives me a formatting screen and then it restarts the device).

**Review by ConnorRigby — APPROVED** [review]

File: lib/nerves_runtime/init.ex [hunk 1]
```diff
@@ -158,4 +159,15 @@ defmodule Nerves.Runtime.Init do
   end
 
   defp validate_mount(s), do: s.mounted
+
+  defp maybe_restart(%{mounted: :mounted, format_performed: true} = s) do
+    # If we formatted and it successfully mounted, we need to restart
+    # before using the partition to prevent errors. This typically
+    # only occurs on first boot
+    Logger.warn("Format and mount successful - restarting...")
```
Comment (fhunleth) [inline lib/nerves_runtime/init.ex:167]:
```suggestion
    Logger.warn("Application data partition successfully formatted - restarting VM...")
```

**Comment by fhunleth** [comment]
I assume what @ConnorRigby is wondering with OTP 24 is how long this takes since unloading code is slow. It seems like it's not that slow in your test, but I wonder if it will be unbearably slow in a larger app.

**Comment by jjcarstens** [comment]
It seems to be anywhere from 12-30 seconds extra in my tests. Its not very noticeable, but there is the chance someone does get an ssh session or serial console up in that and it reboots, which would seem suspect.

**Comment by fhunleth** [comment]
I'm a little concerned about the extra iex prompt and the 12-30 second longer boot time. This is the first thing that people see from Nerves and we get a lot of questions already from the kernel prints when the first time format is detected. 

I am feeling like we should sit on this longer to see if there's a way to improve this.

**Comment by fhunleth** [comment]
I just found out that if I run this on first boot that shell history works again. I looked at the code and Erlang turns off history on an error, but checks that app environment every time to see if the setting changed:

```elixir
Application.put_env(:kernel, :shell_history, :enabled)
```

Perhaps after a format, we run that line. The caveat is what about users who keep shell history off. My best thought so far is to check whether `:shell_history` is set to `:disabled` before switching it to `:enabled`.  Erlang defaults to off if unset. It only looks like Erlang sets it to `:disabled` on errors so presumably it wouldn't be set if the user wants it off or has installed a custom handler.

**Comment by jjcarstens** [comment]
I dug into this a little deeper and tested your logic a bit. It mostly works but does not cover the case of a user setting their own module via `config :kernel, :shell_history, LogModule`

I was hoping to find a way to claim the `:shell_history` value before the shell gets started and potentially wiped out on error but I'm not sure there is an easy way. We could create a `Nerves.IEx.CLI` that is used in the `-user` arg instead of `Elixir.IEx.CLI`. Its sole function would be claim that any `:shell_history` value then call `Elixir.IEx.CLI`. Later here we can pull the stored user value and reassign after reformatting. However, this would only work with projects using the latest `mix nerves.new` generator as it would have to be adjusted there.

We could also read the `ERL_AFLAGS` option to find a module value potentially put there.

It seems like we can cover a lot of cases, but not all of them. So maybe just looking at `Application.get_env :kernel, :shell_history` is sufficient knowing we may lose a user-defined shell history module?

**Comment by jjcarstens** [comment]
So...scratch all that. Turns out a user cannot set base applications in `config.exs`.
They can specify `-kernel shell_history MyLogMod` in `vm.args` so I need to do a little more research there to see if we can find that value as defined

**Comment by jjcarstens** [comment]
K, this PR has been updated to use Franks suggestion. And it will use and `-kernel shell_history` value supplied at startup (either via `ERL_AFLAGS` or `vm.args`) or default to `:enabled` if nothing was previously set by the user

**Review by fhunleth — APPROVED** [review]
Awesome. This feels so much simpler and lower risk than our original ideas.

**Pull request merged**